docs/tutorials/introduction-to-jax.ipynb

```diff
@@ -0,0 +1,646 @@
+    "- **Differentiable Programming:** JAX comes with high-performance automatic differentiation.\n",
+    "\n",
+    "This tutorial is meant to give you a brief introduction to JAX and is structured as follows.\n",
+    "\n",
+    "- [An accelerated NumPy](#an-accelerated-numpy)\n",
+    "- [Useful higher-order functions](#useful-higher-order-functions)\n",
+    "- [PyTrees](#pytrees)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "(an-accelerated-numpy)=\n",
+    "## An accelerated NumPy\n",
+    "\n",
+    "A major appeal of JAX is that its API is largely based on the highly prevalent scientific computing library NumPy.\n",
+    "As a consequence, many programs would run just as well in JAX if we replaced `np` with `jnp`.\n",
+    "Let's look at a few lines of code written in JAX, to illustrate this:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import jax\n",
+    "import jax.numpy as jnp\n",
+    "import numpy as np\n",
+    "\n",
+    "x = jnp.linspace(0, 5, 11)\n",
+    "x"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "unit_circle_arr = jnp.sin(x) ** 2 + jnp.cos(x) ** 2\n",
+    "print(jnp.round(jnp.mean(unit_circle_arr), 7))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Apart from the standard float size of 32 bits, this looks like good old NumPy!\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Purely functional programming and the difference between JAX and NumPy\n",
+    "\n",
+    "To understand the nuances between NumPy and JAX, it is important to note that many of the benefits of plain JAX only really manifest themselves when code is written in a **purely functional** paradigm,\n",
+    "i.e. programs that are constructed by applying and composing pure functions.\n",
+    "\n",
+    "#### Pure functions\n",
+    "\n",
+    "A function is said to be \"pure\" if its **return value is determined solely by its input parameters** and it has **no side effects**.\n",
+    "A side effect is any effect of a function that isn't returned in its output.\n",
+    "Modifying an object passed to the function is an example of a side effect.\n",
+    "Pure functions have distinct properties that are conducive to advanced optimisations,\n",
+    "which is, of course, in line with JAX's primary goal of improving performance.\n",
+    "\n",
+    "#### Array modifications\n",
+    "Let's look at a case NumPy's impure implementation leads to differing behavior between JAX and NumPy:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def sort_array(arr, library_name: str = \"\") -> None:\n",
```

Comment:
For consistency, add data type for `arr`?

Reply:
This is indeed a somewhat awkward inconsistency that I could have easily avoided (`jax.typing.ArrayLike`). Thanks for bringing it to my attention :)